File: src/utils/tracking.ts
import { request as invoke } from './request';

export interface TrackingMetadata {
    screen?: string;
    button?: string | null;
    [key: string]: any;
}

/**
 * Track an event
 * @param eventName - Name of the event (e.g., 'open_app', 'page_view', 'button_click')
 * @param userId - Optional user identifier
 * @param metadata - Additional event metadata
 */
export async function trackEvent(
    eventName: string,
    userId?: string,
    metadata: TrackingMetadata = {}
): Promise<void> {
    try {
        await invoke('track_event', {
            eventName,
            userId: userId || undefined,
            metadata,
        });
    } catch (error) {
        // Silently fail - tracking should not break the app
        console.warn('Failed to track event:', error);
    }
}

/**
 * Get device ID
 */
export async function getDeviceId(): Promise<string> {
    return await invoke<string>('get_tracking_device_id');
}

/**
 * Get machine name
 */
export async function getMachineName(): Promise<string> {
    return await invoke<string>('get_tracking_machine_name');
}

/**
 * Get OS
 */
export async function getOS(): Promise<string> {
    return await invoke<string>('get_tracking_os');
}

// Common event names
export const TrackingEvents = {
    APP_OPEN: 'open_app',
    APP_CLOSE: 'app_close',
    PAGE_VIEW: 'page_view',
    BUTTON_CLICK: 'button_click',
    FEATURE_USE: 'feature_use',
    ACCOUNT_ADD: 'account_add',
    ACCOUNT_SWITCH: 'account_switch',
    PROXY_START: 'proxy_start',
    PROXY_STOP: 'proxy_stop',
    UPDATE_CHECK: 'update_check',
} as const;

File: src-tauri/src/commands/tracking.rs
use crate::modules::tracking;
use serde_json::json;

/// Track an event
#[tauri::command]
pub async fn track_event(
    event_name: String,
    user_id: Option<String>,
    metadata: serde_json::Value,
) -> Result<(), String> {
    let event = tracking::create_event(
        &event_name,
        user_id.as_deref(),
        metadata,
    )?;
    
    // Send tracking event asynchronously (don't block on response)
    tauri::async_runtime::spawn(async move {
        if let Err(e) = tracking::send_tracking_event(event).await {
            tracing::warn!("Failed to send tracking event: {}", e);
        }
    });
    
    Ok(())
}

/// Get device ID
#[tauri::command]
pub fn get_tracking_device_id() -> Result<String, String> {
    tracking::get_device_id()
}

/// Get machine name
#[tauri::command]
pub fn get_tracking_machine_name() -> String {
    tracking::get_machine_name()
}

/// Get OS
#[tauri::command]
pub fn get_tracking_os() -> String {
    tracking::get_os()
}
